certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,240 @@
+import "helpers.spec"
+
+methods {
+    pushFront(bytes32)                    envfree
+    pushBack(bytes32)                     envfree
+    popFront()          returns (bytes32) envfree
+    popBack()           returns (bytes32) envfree
+    clear()                               envfree
+
+    // exposed for FV
+    begin()             returns (int128)  envfree
+    end()               returns (int128)  envfree
+    
+    // view
+    length()            returns (uint256) envfree
+    empty()             returns (bool)    envfree
+    front()             returns (bytes32) envfree
+    back()              returns (bytes32) envfree
+    at_(uint256)        returns (bytes32) envfree // at is a reserved word
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Helpers                                                                                                             │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+
+function absolute_diff(mathint a, mathint b) returns mathint {
+    return a > b ? a - b : b - a; 
+}
+
+function min_int128() returns mathint {
+    return max_uint256 / 2 + 1;
+}
+
+function max_int128() returns mathint {
+    return max_uint256 / 2;
+}
+
+// Could be broken in theory, but not in practice
+function beginLimited(int128 begin) returns bool {
+    return min_int128() < to_mathint(begin); // not inclusive so it can pushFront
+}
+
+// Could be broken in theory, but not in practice
+function endLimited(int128 end) returns bool {
+    return max_int128() > to_mathint(end); // not inclusive so it can pushBack
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: end is larger or equal than begin if values are safely assumed not to overflow.                          │
+| NOTE: This assumption make the proofs unsound, but they can't realistically overflow                                |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant boundariesConsistency()
+    end() >= begin()
+    {
+        preserved {
+            require beginLimited(begin());
+            require endLimited(end());
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: length is end minus begin                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant lengthConsistency()
+    length() == absolute_diff(to_mathint(end()), to_mathint(begin()))
```

do we need the absolute here ? `boundariesConsistency` is `end() >= begin()` ... so it should be good without the absolute value.

I would expect that to work.
```suggestion
    length() == to_mathint(end() - begin())
```

---

You're correct, `absolute_diff` was removed so I'm solving